website/docs/externaldata.md

```diff
@@ -0,0 +1,166 @@
+	--set enableExternalData=true \
+	--set controllerManager.dnsPolicy=ClusterFirst,audit.dnsPolicy=ClusterFirst
+```
+
+> Please note that setting dnsPolicy is a workaround for a bug that is already fixed and will be available in a future release.
+
+### Dev/Test
+
+For dev/test deployments, use `make deploy ENABLE_EXTERNAL_DATA=true`
+
+## Providers
+
+Providers are designed to be in-cluster components that can communicate with external data sources (such as image registries, Active Directory/LDAP directories, etc) and return data in a format that can be processed by Gatekeeper.
+
+Example provider can be found at: https://github.com/open-policy-agent/gatekeeper/tree/master/test/externaldata/dummy-provider
+
+## Provider implementation
+
+Provider is an HTTP server that listens on a port and responds to [ProviderRequest](#ProviderRequest) with [ProviderResponse](#ProviderResponse).
+
+As part of [ProviderResponse](#ProviderResponse), the provider can return a list of items. Each item is a JSON object with the following fields:
+- `Key`: the key that was sent to the provider
+- `Value`: the value that was returned from the provider for that key
+- `Error`: an error message if the provider returned an error for that key
+
+If there is a system error, the provider should return the system error message in the `SystemError` field.
+
+> Recommendation is to keep a timeout such as maximum of 1-2 seconds for the provider to respond.
+
+Example provider implementation: https://github.com/open-policy-agent/gatekeeper/blob/master/test/externaldata/dummy-provider/provider.go
+
+## Gatekeeper Provider custom resource
+
+Provider resource defines the provider and the configuration for it.
+
+```yaml
+apiVersion: externaldata.gatekeeper.sh/v1alpha1
+kind: Provider
+metadata:
+  name: my-provider
+spec:
+  url: http://<service-name>.<namespace>:<port>/<endpoint> # URL to the external data source (e.g., http://my-provider.my-namespace:8090/validate)
+  timeout: <timeout> # timeout value in seconds (e.g., 1)
+```
+
+## Gatekeeper validation
+
+External data adds a [custom OPA built-in function](https://www.openpolicyagent.org/docs/latest/extensions/#custom-built-in-functions-in-go) called `external_data` to Rego. This function is used to query external data providers.
+
+`external_data` is a function that takes a request and returns a response. The request is a JSON object with the following fields:
+- `Provider`: the name of the provider to query
+- `Keys`: the list of keys to send to the provider
+
+```rego
+  # send external data request
+  response := external_data({"provider": "my-provider", "keys": my-list})
+```
+
+Response example: [[`"my-key"`, `"my-value"`, `""`], [`"another-key"`, `42`, `""`], [`"bad-key"`, `""`, `"error message"`]]
+
+> To avoid multiple calls to the provider, recommendation is to batch the keys list and send the request only once.
```

Comment:
```suggestion
> To avoid multiple calls to the same provider, recommendation is to batch the keys list to send a single the request.
```

Reply:
might be good to provide an example of the batch request keys list.

Reply:
added